src/apps/logging/logging.cpp

```diff
@@ -257,6 +259,81 @@ namespace loggingapp
       };
       // SNIPPET_END: log_record_text
 
+      auto& historical_state = context.get_historical_state();
+      auto get_historical = [this, &historical_state](ccf::RequestArgs& args) {
+        const auto params =
+          nlohmann::json::parse(args.rpc_ctx->get_request_body());
+        const auto in = params.get<LoggingGetHistorical::In>();
+
+        // Check that the requested transaction ID is committed
+        {
+          const auto tx_view = consensus->get_view(in.seqno);
+          const auto committed_seqno = consensus->get_committed_seqno();
+          const auto committed_view = consensus->get_view(committed_seqno);
+
+          const auto tx_status = ccf::get_tx_status(
+            in.view, in.seqno, tx_view, committed_view, committed_seqno);
+          if (tx_status != ccf::TxStatus::Committed)
+          {
+            args.rpc_ctx->set_response_status(HTTP_STATUS_BAD_REQUEST);
+            args.rpc_ctx->set_response_header(
+              http::headers::CONTENT_TYPE,
+              http::headervalues::contenttype::TEXT);
+            args.rpc_ctx->set_response_body(fmt::format(
+              "Only committed transactions can be retrieved historically. "
+              "Transaction {}.{} is {}",
+              in.view,
+              in.seqno,
+              ccf::tx_status_to_str(tx_status)));
+            return;
+          }
+        }
+
+        auto historical_store = historical_state.get_store_at(in.seqno);
+        if (historical_store == nullptr)
+        {
+          args.rpc_ctx->set_response_status(HTTP_STATUS_ACCEPTED);
+          static constexpr size_t retry_after_seconds = 3;
+          args.rpc_ctx->set_response_header(
+            http::headers::RETRY_AFTER, retry_after_seconds);
+          args.rpc_ctx->set_response_header(
+            http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+          args.rpc_ctx->set_response_body(fmt::format(
+            "Historical transaction {}.{} is not currently available.",
+            in.view,
+            in.seqno));
+          return;
+        }
+
+        auto* historical_map = historical_store->get(records);
+        if (historical_map == nullptr)
+        {
+          args.rpc_ctx->set_response_status(HTTP_STATUS_INTERNAL_SERVER_ERROR);
+          args.rpc_ctx->set_response_header(
+            http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+          args.rpc_ctx->set_response_body(fmt::format(
+            "Unable to get historical table '{}'.", records.get_name()));
+          return;
+        }
+
+        kv::Tx historical_tx;
+        auto view = historical_tx.get_view(*historical_map);
+        const auto v = view->get(in.id);
+
+        if (v.has_value())
```

Comment:
Just a question. At this point the historical query has been answered, show it be deleted from memory to save EPC?

Reply:
Still an open question. If we delete here, we make some things unnecessarily more expensive (2 users asking for the same version, or the same user asking twice - we have to do a full refetch!). And when we support range queries, its not so obvious we're immediately done with this one.

My gut is that the model should be "supporting historical queries means reserving N bytes of EPC". We'll cull old states so we don't store too many, but we'll keep states around in case they're still useful.